File: frontend/src/mindmapLayout.ts
export type MindMapLayoutNode = {
  id: string;
  width: number;
  height: number;
};

export type MindMapLayoutEdge = {
  source: string;
  target: string;
};

export type MindMapLayoutSide = "root" | "right" | "left";

export type MindMapLayoutPoint = {
  x: number;
  y: number;
};

export type MindMapLayoutBox = {
  left: number;
  top: number;
  right: number;
  bottom: number;
};

export type MindMapLayoutResult = {
  positions: Map<string, MindMapLayoutPoint>;
  boxes: Map<string, MindMapLayoutBox>;
  sideByNode: Map<string, MindMapLayoutSide>;
  rootChildrenRight: string[];
  rootChildrenLeft: string[];
  subtreeExtents: Map<string, number>;
  canvasWidth: number;
  canvasHeight: number;
};

export type MindMapVisibilityResult = {
  visibleNodeIds: string[];
  hiddenCounts: Map<string, number>;
};

export const mindMapLabelMaxLines = ({
  isRoot,
  isBranch,
  hasMedia,
}: {
  isRoot: boolean;
  isBranch: boolean;
  hasMedia: boolean;
}) => {
  if (isRoot) return 8;
  if (hasMedia) return 2;
  return isBranch ? 5 : 6;
};

type OrderedTree = {
  nodeIds: string[];
  children: Map<string, string[]>;
  parent: Map<string, string>;
  depth: Map<string, number>;
  traversal: string[];
};

const unique = (values: string[]) => [...new Set(values)];

const buildOrderedTree = (
  nodeIds: string[],
  edges: MindMapLayoutEdge[],
  rootId: string,
): OrderedTree => {
  const orderedIds = unique(nodeIds);
  const nodeSet = new Set(orderedIds);
  if (!nodeSet.has(rootId)) throw new Error(`布局根节点不存在: ${rootId}`);

  const rawChildren = new Map(
    orderedIds.map((nodeId) => [nodeId, [] as string[]]),
  );
  const rawParent = new Map<string, string>();
  for (const edge of edges) {
    if (
      !nodeSet.has(edge.source) ||
      !nodeSet.has(edge.target) ||
      edge.source === edge.target ||
      edge.target === rootId ||
      rawParent.has(edge.target)
    ) {
      continue;
    }
    rawParent.set(edge.target, edge.source);
    rawChildren.get(edge.source)?.push(edge.target);
  }

  const children = new Map(
    orderedIds.map((nodeId) => [nodeId, [] as string[]]),
  );
  const parent = new Map<string, string>();
  const depth = new Map<string, number>([[rootId, 0]]);
  const traversal = [rootId];
  const visited = new Set<string>([rootId]);

  const walkFrom = (seed: string) => {
    const stack = [seed];
    while (stack.length) {
      const nodeId = stack.pop();
      if (!nodeId) continue;
      const discovered: string[] = [];
      for (const childId of rawChildren.get(nodeId) || []) {
        if (visited.has(childId)) continue;
        visited.add(childId);
        parent.set(childId, nodeId);
        depth.set(childId, (depth.get(nodeId) || 0) + 1);
        children.get(nodeId)?.push(childId);
        traversal.push(childId);
        discovered.push(childId);
      }
      for (let index = discovered.length - 1; index >= 0; index -= 1) {
        stack.push(discovered[index]);
      }
    }
  };

  walkFrom(rootId);
  while (visited.size < orderedIds.length) {
    const unvisited = orderedIds.filter((nodeId) => !visited.has(nodeId));
    const unvisitedSet = new Set(unvisited);
    const seed =
      unvisited.find((nodeId) => {
        const rawParentId = rawParent.get(nodeId);
        return !rawParentId || !unvisitedSet.has(rawParentId);
      }) || unvisited[0];
    visited.add(seed);
    parent.set(seed, rootId);
    depth.set(seed, 1);
    children.get(rootId)?.push(seed);
    traversal.push(seed);
    walkFrom(seed);
  }

  return {
    nodeIds: orderedIds,
    children,
    parent,
    depth,
    traversal,
  };
};

const stackCenters = (
  nodeIds: string[],
  extents: Map<string, number>,
  verticalGap: number,
) => {
  const centers = new Map<string, number>();
  if (!nodeIds.length) return centers;
  const total =
    nodeIds.reduce((sum, nodeId) => sum + (extents.get(nodeId) || 0), 0) +
    verticalGap * (nodeIds.length - 1);
  let cursor = -total / 2;
  for (const nodeId of nodeIds) {
    const extent = extents.get(nodeId) || 0;
    centers.set(nodeId, cursor + extent / 2);
    cursor += extent + verticalGap;
  }
  return centers;
};

const splitRootChildren = (
  rootChildren: string[],
  extents: Map<string, number>,
  verticalGap: number,
  rightRatio: number,
) => {
  if (!rootChildren.length) return { right: [] as string[], left: [] as string[] };
  const totalHeight =
    rootChildren.reduce(
      (sum, nodeId) => sum + (extents.get(nodeId) || 0),
      0,
    ) +
    verticalGap * (rootChildren.length - 1);
  const rightBudget = totalHeight * rightRatio;
  const right: string[] = [];
  const left: string[] = [];
  let usedHeight = 0;
  let switched = false;
  for (const nodeId of rootChildren) {
    const addition =
      (extents.get(nodeId) || 0) + (right.length ? verticalGap : 0);
    if (!switched && (!right.length || usedHeight + addition <= rightBudget)) {
      right.push(nodeId);
      usedHeight += addition;
    } else {
      switched = true;
      left.push(nodeId);
    }
  }
  return { right, left };
};

export const computeMindMapLayout = ({
  nodes,
  edges,
  rootId,
  rightRatio = 0.62,
  verticalGap = 32,
  horizontalGap = 72,
  margin = 72,
  titleSpace = 84,
  minimumCanvasWidth = 1200,
  minimumCanvasHeight = 720,
}: {
  nodes: MindMapLayoutNode[];
  edges: MindMapLayoutEdge[];
  rootId: string;
  rightRatio?: number;
  verticalGap?: number;
  horizontalGap?: number;
  margin?: number;
  titleSpace?: number;
  minimumCanvasWidth?: number;
  minimumCanvasHeight?: number;
}): MindMapLayoutResult => {
  if (rightRatio <= 0 || rightRatio > 1) {
    throw new Error("rightRatio 必须在 (0, 1] 内");
  }
  if (verticalGap < 24 || horizontalGap < 24) {
    throw new Error("布局节点安全间距不能小于 24px");
  }

  const tree = buildOrderedTree(
    nodes.map((node) => node.id),
    edges,
    rootId,
  );
  const sizeById = new Map(nodes.map((node) => [node.id, node]));
  const subtreeExtents = new Map<string, number>();
  for (const nodeId of [...tree.traversal].reverse()) {
    const node = sizeById.get(nodeId);
    if (!node || node.width <= 0 || node.height <= 0) {
      throw new Error(`节点缺少有效尺寸: ${nodeId}`);
    }
    const childIds = tree.children.get(nodeId) || [];
    const childrenHeight =
      childIds.reduce(
        (sum, childId) => sum + (subtreeExtents.get(childId) || 0),
        0,
      ) + verticalGap * Math.max(childIds.length - 1, 0);
    subtreeExtents.set(nodeId, Math.max(node.height, childrenHeight));
  }

  const maxWidthByDepth = new Map<number, number>();
  let maxDepth = 0;
  for (const nodeId of tree.nodeIds) {
    const nodeDepth = tree.depth.get(nodeId) || 0;
    const node = sizeById.get(nodeId);
    if (!node) throw new Error(`节点缺少有效尺寸: ${nodeId}`);
    maxDepth = Math.max(maxDepth, nodeDepth);
    maxWidthByDepth.set(
      nodeDepth,
      Math.max(maxWidthByDepth.get(nodeDepth) || 0, node.width),
    );
  }
  const columnX = new Map<number, number>([[0, 0]]);
  for (let nodeDepth = 1; nodeDepth <= maxDepth; nodeDepth += 1) {
    columnX.set(
      nodeDepth,
      (columnX.get(nodeDepth - 1) || 0) +
        (maxWidthByDepth.get(nodeDepth - 1) || 1) / 2 +
        horizontalGap +
        (maxWidthByDepth.get(nodeDepth) || 1) / 2,
    );
  }

  const rootChildren = tree.children.get(rootId) || [];
  const { right, left } = splitRootChildren(
    rootChildren,
    subtreeExtents,
    verticalGap,
    rightRatio,
  );
  const rawPositions = new Map<string, MindMapLayoutPoint>([
    [rootId, { x: 0, y: 0 }],
  ]);
  const sideByNode = new Map<string, MindMapLayoutSide>([[rootId, "root"]]);

  const placeSubtree = (
    seed: string,
    seedY: number,
    side: Exclude<MindMapLayoutSide, "root">,
  ) => {
    const stack: Array<[string, number]> = [[seed, seedY]];
    while (stack.length) {
      const current = stack.pop();
      if (!current) continue;
      const [nodeId, centerY] = current;
      const nodeDepth = tree.depth.get(nodeId) || 0;
      rawPositions.set(nodeId, {
        x: (side === "right" ? 1 : -1) * (columnX.get(nodeDepth) || 0),
        y: centerY,
      });
      sideByNode.set(nodeId, side);
      const childIds = tree.children.get(nodeId) || [];
      const childCenters = stackCenters(childIds, subtreeExtents, verticalGap);
      const descendants = childIds.map(
        (childId) =>
          [childId, centerY + (childCenters.get(childId) || 0)] as [
            string,
            number,
          ],
      );
      for (let index = descendants.length - 1; index >= 0; index -= 1) {
        stack.push(descendants[index]);
      }
    }
  };

  for (const [branchId, centerY] of stackCenters(
    right,
    subtreeExtents,
    verticalGap,
  )) {
    placeSubtree(branchId, centerY, "right");
  }
  for (const [branchId, centerY] of stackCenters(
    left,
    subtreeExtents,
    verticalGap,
  )) {
    placeSubtree(branchId, centerY, "left");
  }

  const rawBoxes = new Map<string, MindMapLayoutBox>();
  for (const [nodeId, point] of rawPositions) {
    const node = sizeById.get(nodeId);
    if (!node) continue;
    rawBoxes.set(nodeId, {
      left: point.x - node.width / 2,
      top: point.y - node.height / 2,
      right: point.x + node.width / 2,
      bottom: point.y + node.height / 2,
    });
  }
  const boxValues = [...rawBoxes.values()];
  const minX = Math.min(...boxValues.map((box) => box.left));
  const maxX = Math.max(...boxValues.map((box) => box.right));
  const minY = Math.min(...boxValues.map((box) => box.top));
  const maxY = Math.max(...boxValues.map((box) => box.bottom));
  const contentWidth = maxX - minX;
  const contentHeight = maxY - minY;
  const canvasWidth = Math.max(
    minimumCanvasWidth,
    Math.ceil(contentWidth + margin * 2),
  );
  const canvasHeight = Math.max(
    minimumCanvasHeight,
    Math.ceil(contentHeight + margin * 2 + titleSpace),
  );
  const offsetX = (canvasWidth - contentWidth) / 2 - minX;
  const offsetY =
    titleSpace + (canvasHeight - titleSpace - contentHeight) / 2 - minY;
  const positions = new Map(
    [...rawPositions].map(([nodeId, point]) => [
      nodeId,
      { x: point.x + offsetX, y: point.y + offsetY },
    ]),
  );
  const boxes = new Map(
    [...rawBoxes].map(([nodeId, box]) => [
      nodeId,
      {
        left: box.left + offsetX,
        top: box.top + offsetY,
        right: box.right + offsetX,
        bottom: box.bottom + offsetY,
      },
    ]),
  );
  return {
    positions,
    boxes,
    sideByNode,
    rootChildrenRight: right,
    rootChildrenLeft: left,
    subtreeExtents,
    canvasWidth,
    canvasHeight,
  };
};

export const findMindMapSpacingViolations = (
  layout: MindMapLayoutResult,
  minimumGap = 24,
) => {
  const nodeIds = [...layout.positions.keys()];
  const violations: Array<[string, string]> = [];
  for (let leftIndex = 0; leftIndex < nodeIds.length; leftIndex += 1) {
    const leftId = nodeIds[leftIndex];
    const leftBox = layout.boxes.get(leftId);
    if (!leftBox) continue;
    for (const rightId of nodeIds.slice(leftIndex + 1)) {
      const rightBox = layout.boxes.get(rightId);
      if (!rightBox) continue;
      const horizontalGap = Math.max(
        rightBox.left - leftBox.right,
        leftBox.left - rightBox.right,
        0,
      );
      const verticalGap = Math.max(
        rightBox.top - leftBox.bottom,
        leftBox.top - rightBox.bottom,
        0,
      );
      if (horizontalGap < minimumGap && verticalGap < minimumGap) {
        violations.push([leftId, rightId]);
      }
    }
  }
  return violations;
};

export const collapseMindMapToBudget = ({
  nodeIds,
  edges,
  rootId,
  maxVisible = 120,
}: {
  nodeIds: string[];
  edges: MindMapLayoutEdge[];
  rootId: string;
  maxVisible?: number;
}): MindMapVisibilityResult => {
  if (maxVisible < 1) throw new Error("maxVisible 必须至少为 1");
  const tree = buildOrderedTree(nodeIds, edges, rootId);
  const visible = new Set<string>([rootId]);
  const pending = [rootId];
  let cursor = 0;
  while (cursor < pending.length && visible.size < maxVisible) {
    const parentId = pending[cursor];
    cursor += 1;
    for (const childId of tree.children.get(parentId) || []) {
      if (visible.size >= maxVisible) break;
      visible.add(childId);
      pending.push(childId);
    }
  }

  const hiddenCounts = new Map<string, number>();
  for (const nodeId of tree.nodeIds) {
    if (visible.has(nodeId)) continue;
    let ancestorId = tree.parent.get(nodeId) || rootId;
    while (!visible.has(ancestorId)) {
      ancestorId = tree.parent.get(ancestorId) || rootId;
    }
    hiddenCounts.set(ancestorId, (hiddenCounts.get(ancestorId) || 0) + 1);
  }
  return {
    visibleNodeIds: tree.nodeIds.filter((nodeId) => visible.has(nodeId)),
    hiddenCounts,
  };
};

export const wrapMindMapLabel = (
  value: string,
  unitsPerLine: number,
  maxLines: number,
) => {
  const lines: string[] = [];
  let current = "";
  let currentUnits = 0;
  const characters = Array.from(value.trim());
  let consumed = 0;
  for (const character of characters) {
    const units = /^[\x00-\x7F]$/.test(character) ? 0.55 : 1;
    if (current && currentUnits + units > unitsPerLine) {
      lines.push(current);
      current = "";
      currentUnits = 0;
      if (lines.length === maxLines) break;
    }
    current += character;
    currentUnits += units;
    consumed += 1;
  }
  if (current && lines.length < maxLines) lines.push(current);
  if (consumed < characters.length && lines.length > 0) {
    const last = lines.length - 1;
    lines[last] = `${lines[last].slice(0, -1)}…`;
  }
  return lines.join("\n") || value;
};

File: backend/app/mindmap_layout.py
from __future__ import annotations

from collections import deque
from dataclasses import dataclass
from math import ceil, isfinite, sqrt
from typing import Iterable, Literal, Mapping, Sequence


LayoutSide = Literal["root", "right", "left"]


@dataclass(frozen=True, slots=True)
class NodeSize:
    width: float
    height: float

    def __post_init__(self) -> None:
        if (
            not isfinite(self.width)
            or not isfinite(self.height)
            or self.width <= 0
            or self.height <= 0
        ):
            raise ValueError("node width and height must be finite positive values")


@dataclass(frozen=True, slots=True)
class LayoutPoint:
    x: float
    y: float


@dataclass(frozen=True, slots=True)
class LayoutBox:
    left: float
    top: float
    right: float
    bottom: float


@dataclass(frozen=True, slots=True)
class LayoutResult:
    positions: dict[str, LayoutPoint]
    sizes: dict[str, NodeSize]
    boxes: dict[str, LayoutBox]
    side_by_node: dict[str, LayoutSide]
    root_children_right: tuple[str, ...]
    root_children_left: tuple[str, ...]
    subtree_extents: dict[str, float]
    canvas_width: int
    canvas_height: int
    content_bounds: LayoutBox


@dataclass(frozen=True, slots=True)
class VisibilityResult:
    visible_node_ids: tuple[str, ...]
    hidden_counts: dict[str, int]


@dataclass(frozen=True, slots=True)
class RasterPlan:
    width: int
    height: int
    scale: float


@dataclass(slots=True)
class _OrderedTree:
    node_ids: list[str]
    children: dict[str, list[str]]
    parent: dict[str, str]
    depth: dict[str, int]
    traversal: list[str]


def _deduplicate(values: Iterable[str]) -> list[str]:
    seen: set[str] = set()
    ordered: list[str] = []
    for value in values:
        if value in seen:
            continue
        seen.add(value)
        ordered.append(value)
    return ordered


def _build_ordered_tree(
    node_ids: Sequence[str],
    edges: Sequence[tuple[str, str]],
    root_id: str,
) -> _OrderedTree:
    ordered_ids = _deduplicate(node_ids)
    node_set = set(ordered_ids)
    if root_id not in node_set:
        raise ValueError(f"root node {root_id!r} is not present")

    raw_children = {node_id: [] for node_id in ordered_ids}
    raw_parent: dict[str, str] = {}
    for source, target in edges:
        if (
            source not in node_set
            or target not in node_set
            or source == target
            or target == root_id
            or target in raw_parent
        ):
            continue
        raw_parent[target] = source
        raw_children[source].append(target)

    children = {node_id: [] for node_id in ordered_ids}
    parent: dict[str, str] = {}
    depth = {root_id: 0}
    traversal = [root_id]
    visited = {root_id}

    def walk_from(seed: str) -> None:
        stack = [seed]
        while stack:
            node_id = stack.pop()
            discovered: list[str] = []
            for child_id in raw_children[node_id]:
                if child_id in visited:
                    continue
                visited.add(child_id)
                parent[child_id] = node_id
                depth[child_id] = depth[node_id] + 1
                children[node_id].append(child_id)
                traversal.append(child_id)
                discovered.append(child_id)
            stack.extend(reversed(discovered))

    walk_from(root_id)

    # A valid solver result is one rooted tree. Keeping this deterministic
    # fallback makes layout/export resilient to old or partially reviewed
    # graph versions without silently dropping disconnected nodes.
    while len(visited) < len(ordered_ids):
        unvisited = set(ordered_ids) - visited
        seed = next(
            (
                node_id
                for node_id in ordered_ids
                if node_id in unvisited
                and raw_parent.get(node_id) not in unvisited
            ),
            next(node_id for node_id in ordered_ids if node_id in unvisited),
        )
        visited.add(seed)
        parent[seed] = root_id
        depth[seed] = 1
        children[root_id].append(seed)
        traversal.append(seed)
        walk_from(seed)

    return _OrderedTree(
        node_ids=ordered_ids,
        children=children,
        parent=parent,
        depth=depth,
        traversal=traversal,
    )


def _stack_centers(
    node_ids: Sequence[str],
    extents: Mapping[str, float],
    vertical_gap: float,
) -> dict[str, float]:
    if not node_ids:
        return {}
    total = sum(extents[node_id] for node_id in node_ids)
    total += vertical_gap * (len(node_ids) - 1)
    cursor = -total / 2
    centers: dict[str, float] = {}
    for node_id in node_ids:
        extent = extents[node_id]
        centers[node_id] = cursor + extent / 2
        cursor += extent + vertical_gap
    return centers


def _split_root_children(
    root_children: Sequence[str],
    extents: Mapping[str, float],
    vertical_gap: float,
    right_ratio: float,
) -> tuple[list[str], list[str]]:
    if not root_children:
        return [], []
    total_height = sum(extents[node_id] for node_id in root_children)
    total_height += vertical_gap * (len(root_children) - 1)
    right_budget = total_height * right_ratio
    right: list[str] = []
    left: list[str] = []
    used_height = 0.0
    switched = False
    for node_id in root_children:
        addition = extents[node_id] + (vertical_gap if right else 0)
        if not switched and (
            not right or used_height + addition <= right_budget
        ):
            right.append(node_id)
            used_height += addition
            continue
        switched = True
        left.append(node_id)
    return right, left


def compute_mindmap_layout(
    *,
    node_ids: Sequence[str],
    edges: Sequence[tuple[str, str]],
    root_id: str,
    sizes: Mapping[str, NodeSize],
    right_ratio: float = 0.62,
    vertical_gap: float = 32,
    horizontal_gap: float = 72,
    margin: float = 72,
    title_space: float = 84,
    minimum_canvas_width: int = 1200,
    minimum_canvas_height: int = 720,
) -> LayoutResult:
    """Lay out a mind map as a deterministic, right-first, two-sided tree.

    The algorithm uses the supplied final node dimensions. It computes every
    subtree's vertical extent bottom-up, assigns an ordered prefix of root
    branches to the right, and moves the first overflowing branch plus all
    following branches to the left.
    """

    if not 0 < right_ratio <= 1:
        raise ValueError("right_ratio must be in (0, 1]")
    if vertical_gap < 24:
        raise ValueError("vertical_gap must be at least the 24 px safety gap")
    if horizontal_gap < 24:
        raise ValueError("horizontal_gap must be at least the 24 px safety gap")
    if margin < 0 or title_space < 0:
        raise ValueError("margin and title_space cannot be negative")

    tree = _build_ordered_tree(node_ids, edges, root_id)
    missing_sizes = [
        node_id for node_id in tree.node_ids if node_id not in sizes
    ]
    if missing_sizes:
        raise ValueError(f"missing node sizes for: {', '.join(missing_sizes[:5])}")
    normalized_sizes = {
        node_id: sizes[node_id]
        for node_id in tree.node_ids
    }

    subtree_extents: dict[str, float] = {}
    for node_id in reversed(tree.traversal):
        child_ids = tree.children[node_id]
        children_height = sum(
            subtree_extents[child_id] for child_id in child_ids
        )
        if child_ids:
            children_height += vertical_gap * (len(child_ids) - 1)
        subtree_extents[node_id] = max(
            normalized_sizes[node_id].height,
            children_height,
        )

    max_width_by_depth: dict[int, float] = {}
    max_depth = 0
    for node_id in tree.node_ids:
        node_depth = tree.depth[node_id]
        max_depth = max(max_depth, node_depth)
        max_width_by_depth[node_depth] = max(
            max_width_by_depth.get(node_depth, 0),
            normalized_sizes[node_id].width,
        )

    column_x = {0: 0.0}
    for node_depth in range(1, max_depth + 1):
        previous_width = max_width_by_depth.get(node_depth - 1, 1)
        current_width = max_width_by_depth.get(node_depth, 1)
        column_x[node_depth] = (
            column_x[node_depth - 1]
            + previous_width / 2
            + horizontal_gap
            + current_width / 2
        )

    root_children = tree.children[root_id]
    right_children, left_children = _split_root_children(
        root_children,
        subtree_extents,
        vertical_gap,
        right_ratio,
    )
    raw_positions = {root_id: LayoutPoint(0, 0)}
    side_by_node: dict[str, LayoutSide] = {root_id: "root"}

    def place_subtree(seed: str, seed_y: float, side: LayoutSide) -> None:
        stack: list[tuple[str, float]] = [(seed, seed_y)]
        while stack:
            node_id, center_y = stack.pop()
            direction = 1 if side == "right" else -1
            raw_positions[node_id] = LayoutPoint(
                direction * column_x[tree.depth[node_id]],
                center_y,
            )
            side_by_node[node_id] = side
            child_ids = tree.children[node_id]
            if not child_ids:
                continue
            child_centers = _stack_centers(
                child_ids,
                subtree_extents,
                vertical_gap,
            )
            descendants = [
                (child_id, center_y + child_centers[child_id])
                for child_id in child_ids
            ]
            stack.extend(reversed(descendants))

    for branch_id, center_y in _stack_centers(
        right_children,
        subtree_extents,
        vertical_gap,
    ).items():
        place_subtree(branch_id, center_y, "right")
    for branch_id, center_y in _stack_centers(
        left_children,
        subtree_extents,
        vertical_gap,
    ).items():
        place_subtree(branch_id, center_y, "left")

    raw_boxes: dict[str, LayoutBox] = {}
    for node_id, point in raw_positions.items():
        size = normalized_sizes[node_id]
        raw_boxes[node_id] = LayoutBox(
            left=point.x - size.width / 2,
            top=point.y - size.height / 2,
            right=point.x + size.width / 2,
            bottom=point.y + size.height / 2,
        )

    min_x = min(box.left for box in raw_boxes.values())
    max_x = max(box.right for box in raw_boxes.values())
    min_y = min(box.top for box in raw_boxes.values())
    max_y = max(box.bottom for box in raw_boxes.values())
    content_width = max_x - min_x
    content_height = max_y - min_y
    canvas_width = max(
        minimum_canvas_width,
        ceil(content_width + margin * 2),
    )
    canvas_height = max(
        minimum_canvas_height,
        ceil(content_height + margin * 2 + title_space),
    )
    offset_x = (canvas_width - content_width) / 2 - min_x
    drawable_height = canvas_height - title_space
    offset_y = title_space + (drawable_height - content_height) / 2 - min_y

    positions = {
        node_id: LayoutPoint(
            x=point.x + offset_x,
            y=point.y + offset_y,
        )
        for node_id, point in raw_positions.items()
    }
    boxes = {
        node_id: LayoutBox(
            left=box.left + offset_x,
            top=box.top + offset_y,
            right=box.right + offset_x,
            bottom=box.bottom + offset_y,
        )
        for node_id, box in raw_boxes.items()
    }
    content_bounds = LayoutBox(
        left=min(box.left for box in boxes.values()),
        top=min(box.top for box in boxes.values()),
        right=max(box.right for box in boxes.values()),
        bottom=max(box.bottom for box in boxes.values()),
    )
    return LayoutResult(
        positions=positions,
        sizes=normalized_sizes,
        boxes=boxes,
        side_by_node=side_by_node,
        root_children_right=tuple(right_children),
        root_children_left=tuple(left_children),
        subtree_extents=subtree_extents,
        canvas_width=canvas_width,
        canvas_height=canvas_height,
        content_bounds=content_bounds,
    )


def find_spacing_violations(
    layout: LayoutResult,
    *,
    minimum_gap: float = 24,
) -> list[tuple[str, str]]:
    """Return box pairs whose AABBs do not have the requested safe gap."""

    if minimum_gap < 0:
        raise ValueError("minimum_gap cannot be negative")
    node_ids = list(layout.positions)
    violations: list[tuple[str, str]] = []
    for left_index, left_id in enumerate(node_ids):
        left_box = layout.boxes[left_id]
        for right_id in node_ids[left_index + 1 :]:
            right_box = layout.boxes[right_id]
            horizontal_gap = max(
                right_box.left - left_box.right,
                left_box.left - right_box.right,
                0,
            )
            vertical_gap = max(
                right_box.top - left_box.bottom,
                left_box.top - right_box.bottom,
                0,
            )
            if horizontal_gap < minimum_gap and vertical_gap < minimum_gap:
                violations.append((left_id, right_id))
    return violations


def collapse_tree_to_budget(
    *,
    node_ids: Sequence[str],
    edges: Sequence[tuple[str, str]],
    root_id: str,
    max_visible: int = 120,
) -> VisibilityResult:
    """Select an ancestor-closed overview and collapse remaining subtrees."""

    if max_visible < 1:
        raise ValueError("max_visible must be at least 1")
    tree = _build_ordered_tree(node_ids, edges, root_id)
    visible = {root_id}
    pending: deque[str] = deque([root_id])
    while pending and len(visible) < max_visible:
        parent_id = pending.popleft()
        for child_id in tree.children[parent_id]:
            if len(visible) >= max_visible:
                break
            visible.add(child_id)
            pending.append(child_id)

    hidden_counts: dict[str, int] = {}
    for node_id in tree.node_ids:
        if node_id in visible:
            continue
        ancestor_id = tree.parent.get(node_id, root_id)
        while ancestor_id not in visible:
            ancestor_id = tree.parent.get(ancestor_id, root_id)
        hidden_counts[ancestor_id] = hidden_counts.get(ancestor_id, 0) + 1

    return VisibilityResult(
        visible_node_ids=tuple(
            node_id for node_id in tree.node_ids if node_id in visible
        ),
        hidden_counts=hidden_counts,
    )


def plan_raster_size(
    *,
    canvas_width: int | float,
    canvas_height: int | float,
    max_pixels: int = 16_000_000,
    max_dimension: int = 8_192,
) -> RasterPlan:
    """Plan a bounded raster without ever allocating the logical full canvas."""

    if canvas_width <= 0 or canvas_height <= 0:
        raise ValueError("canvas dimensions must be positive")
    if max_pixels < 1 or max_dimension < 1:
        raise ValueError("raster limits must be positive")

    scale = min(
        1.0,
        max_dimension / canvas_width,
        max_dimension / canvas_height,
        sqrt(max_pixels / (canvas_width * canvas_height)),
    )
    while True:
        width = max(1, ceil(canvas_width * scale))
        height = max(1, ceil(canvas_height * scale))
        if (
            width * height <= max_pixels
            and width <= max_dimension
            and height <= max_dimension
        ):
            break
        scale *= 0.999
    return RasterPlan(width=width, height=height, scale=scale)
